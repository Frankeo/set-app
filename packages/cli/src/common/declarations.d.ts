declare module "git-file-downloader";
