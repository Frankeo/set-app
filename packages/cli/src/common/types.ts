export interface PackageJSON {
  version: string;
  scripts: ScriptCommands;
  dependencies: Dependencies;
  devDependencies: Dependencies;
}

export interface ScriptCommands {
  lint?: string;
  format?: string;
  "check-format"?: string;
  dev?: string;
  build?: string;
  preview?: string;
  typecheck?: string;
  test?: string;
  "test:watch"?: string;
  coverage?: string;
  postinstall?: string;
  "pre-commit"?: string;
  prepack?: string;
  postpack?: string;
  ci?: string;
  "build-ci"?: string;
}

interface ESLintRules {
  [K: string]: number;
}

export interface Dependencies {
  [K: string]: string;
}

export interface ESLintRC {
  extends: string[];
  plugins: string[];
  parser: string;
  parserOptions: {
    ecmaVersion: number;
    project: string;
    sourceType: string;
    ecmaFeatures?: {
      jsx: boolean;
    };
  };
  rules?: ESLintRules;
  env: {
    es6: boolean;
    browser: boolean;
    node: boolean;
  };
  settings: {
    react?: {
      version: string;
    };
    "import/resolver": {
      typescript: {
        alwaysTryTypes: boolean;
      };
    };
    "import/parsers": {
      "@typescript-eslint/parser": string[];
    };
  };
}

export enum ExampleOptions {
  REACT = "REACT",
  CONSOLE = "CONSOLE",
  REDUX = "REACT-REDUX",
}

export interface CliArguments {
  desc?: string;
  type: ExampleOptions;
  github: boolean;
}

export interface NetlifyConfig {
  userId: string;
  users: {
    [user: string]: {
      auth: {
        token: string;
      };
    };
  };
}

export interface NetlifySiteInfo {
  site_id: string;
}
