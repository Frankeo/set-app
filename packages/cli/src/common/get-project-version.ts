import { join } from "path";
import { fs } from "zx";
import { PackageJSON } from "./types.js";
import { dirName } from "../get-dir-name.js";

export const getProjectVersion = () => {
  const packageJSON: PackageJSON = fs.readJSONSync(
    join(dirName(), "..", "package.json")
  );
  return packageJSON.version;
};
