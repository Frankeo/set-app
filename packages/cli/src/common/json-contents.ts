import { ESLintRC } from "./types.js";

export const esLintRCBasic: ESLintRC = {
  extends: [
    "eslint:recommended",
    "plugin:@typescript-eslint/recommended",
    "prettier",
  ],
  plugins: ["@typescript-eslint"],
  parser: "@typescript-eslint/parser",
  parserOptions: {
    project: "./tsconfig.json",
    ecmaVersion: 2022,
    sourceType: "module",
  },
  env: {
    es6: true,
    browser: true,
    node: true,
  },
  settings: {
    "import/resolver": {
      typescript: {
        alwaysTryTypes: true,
      },
    },
    "import/parsers": {
      "@typescript-eslint/parser": ["ts"],
    },
  },
};

export const esLintRCReact: ESLintRC = {
  extends: [
    "eslint:recommended",
    "plugin:import/errors",
    "plugin:react/recommended",
    "plugin:jsx-a11y/recommended",
    "plugin:react-hooks/recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:@typescript-eslint/recommended-requiring-type-checking",
    "prettier",
  ],
  rules: {
    "react/prop-types": 0,
    "react/react-in-jsx-scope": 0,
    "@typescript-eslint/no-empty-function": 0,
  },
  plugins: ["react", "import", "jsx-a11y", "@typescript-eslint"],
  parser: "@typescript-eslint/parser",
  parserOptions: {
    project: "./tsconfig.json",
    ecmaVersion: 2022,
    sourceType: "module",
    ecmaFeatures: {
      jsx: true,
    },
  },
  env: {
    es6: true,
    browser: true,
    node: true,
  },
  settings: {
    react: {
      version: "detect",
    },
    "import/parsers": {
      "@typescript-eslint/parser": [".ts", ".tsx"],
    },
    "import/resolver": {
      typescript: {
        alwaysTryTypes: true,
      },
    },
  },
};

export const tsConfigBasic = {
  compilerOptions: {
    target: "ES2022",
    jsx: "react-jsx",
    module: "ES2022",
    moduleResolution: "node",
    esModuleInterop: true,
    forceConsistentCasingInFileNames: true,
    strict: true,
    skipLibCheck: true,
  },
};

export const devContainerConfig = {
  name: "Debian",
  image: "mcr.microsoft.com/devcontainers/base:bullseye",
  features: {
    "ghcr.io/devcontainers/features/common-utils:1": {
      installZsh: true,
      installOhMyZsh: true,
      upgradePackages: true,
    },
    "ghcr.io/devcontainers/features/node:1": {},
  },
  settings: {
    "terminal.integrated.defaultProfile.linux": "zsh",
    "terminal.integrated.profiles.linux": {
      zsh: {
        path: "/bin/zsh",
      },
    },
  },
  extensions: [
    "zixuanchen.vitest-explorer",
    "esbenp.prettier-vscode",
    "dbaeumer.vscode-eslint",
    "christian-kohler.npm-intellisense",
    "tal7aouy.icons",
    "dsznajder.es7-react-js-snippets",
  ],
};
