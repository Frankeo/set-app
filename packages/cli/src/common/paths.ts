import { join, dirname } from "path";
import { dirName } from "../get-dir-name.js";

import { ExampleOptions } from "./types.js";

export const getProjectPath = (name: string) => join(process.cwd(), name);

const getToolDirectory = () => dirname(process.argv[1]);

export const getCopyFolder = (name: string) => join(getToolDirectory(), name);

export const getPackageJson = (name: string) =>
  join(getProjectPath(name), "package.json");

export const getSrcPath = (name: string) => join(getProjectPath(name), "src");

export const getIndexHtml = (name: string) =>
  join(getSrcPath(name), "index.html");

export const getEslintRc = (name: string) =>
  join(getProjectPath(name), ".eslintrc.json");

export const getProdEnv = () => join(dirName(), ".env");

export const getExample = (exampleName: ExampleOptions) => {
  const fullExampleName = `EXAMPLE_${exampleName.toUpperCase()}`;
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const examplePath = process.env[fullExampleName]!;
  return join(dirName(), examplePath);
};

export const getExampleSrc = (exampleName: ExampleOptions): string => {
  const exampleSrc = getExample(exampleName);
  return join(exampleSrc, process.env.NODE_ENV ? "src" : "examples-src");
};

export const getDevContainer = (name: string): string => {
  const projectPath = getProjectPath(name);
  return join(projectPath, ".devcontainer");
};

export const getDevContainerJson = (name: string): string => {
  const devContainer = getDevContainer(name);
  return join(devContainer, "devcontainer.json");
};
