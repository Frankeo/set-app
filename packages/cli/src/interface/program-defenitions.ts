import { program } from "commander";
import { createProjectAction } from "./create-project-action.js";
import { getProjectVersion } from "./get-project-version.js";

export const createTool = () =>
  program
    .name("set-app")
    .description("Create and setup node projects")
    .argument("<name>", "project name")
    .option(
      "-t, --type <value>",
      "Choose: React, React-Redux or Console",
      "React"
    )
    .option("--desc <value>", "Add a description to your new project")
    .option("--no-github", "Not generate github public repository")
    .version(getProjectVersion())
    .action(createProjectAction);
