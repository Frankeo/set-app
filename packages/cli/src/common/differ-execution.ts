import { $ } from "zx";
import { getProjectPath } from "./paths.js";

export const devDependencies: string[] = [];
export const dependencies: string[] = [];

const removeBadSustitution = (str: string) =>
  str.replace(new RegExp(/\$'[^']*'/, "g"), (value) =>
    value.replace(new RegExp(/(\$')|(')/, "g"), "")
  );

export const executeDependencies = async (projectName: string) => {
  $.quote = removeBadSustitution;
  const projectPath = getProjectPath(projectName);
  const devDependencyCommand = `yarn add -D ${devDependencies.join(" ")}`;
  const dependencyCommand = `yarn add ${dependencies.join(" ")}`;
  return $`cd ${projectPath} ; ${dependencyCommand} ; ${devDependencyCommand}`;
};
