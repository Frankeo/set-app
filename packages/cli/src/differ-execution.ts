import { dirname } from "path";
import { fileURLToPath } from "url";
import { $ } from "zx";
import { getProjectPath } from "./common/paths.js";
import { updatePackageJsonDependencies } from "./common/update-package-json-script.js";
import { generateLoader, getSuccessMessage } from "./interface/messages.js";

export const executeDependencies = async (projectName: string) => {
  const projectPath = getProjectPath(projectName);
  updatePackageJsonDependencies(projectName);
  const loader = generateLoader("Installing all dependencies");
  await $`cd ${projectPath} ; yarn install`;
  loader.stop();
  getSuccessMessage("installed", "All dependencies");
};

// // Don't move from this file, needs to take the reference
export const dirName = () => dirname(fileURLToPath(import.meta.url));
