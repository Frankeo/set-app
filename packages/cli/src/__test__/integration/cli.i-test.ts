import { expect, test, describe } from "vitest";
import { createTool } from "../../interface/program-defenitions";
import { hashElement } from "folder-hash";
import { dirName } from "../../differ-execution";
import { join } from "path";
import { getProjectPath } from "../../common/paths";
import { rimrafSync } from "rimraf";
import { PackageJSON } from "../../common/types";
import { fs } from "zx";

describe("SetApp cli tool", () => {
  test.each(["-h", "--help"])(
    "when specify %s then show help and exit",
    (arg: string) => {
      const program = createTool()
        .exitOverride()
        .configureOutput({
          writeOut(str) {
            expect(str).toMatchSnapshot();
          },
        });
      expect(() => program.parse(["node", "cli", arg])).toThrow("(outputHelp)");
    }
  );

  test.each(["REACT", "REACT-REDUX"])(
    "when specify projectName then create a default %s project with that name",
    async (type: string) => {
      const projectName = process.env[`CI_TEST_${type}`]!;
      const options = {
        folders: { exclude: [".*", "node_modules", "dist"] },
        files: {
          include: [
            "*.snap",
            "*.jsx",
            "*.ts",
            "*.tsx",
            "*.css",
            "*.html",
            ".prettierrc",
            "LICENSE",
            ".gitignore",
            "README.md",
            ".eslintrc.json",
            "tsconfig.json",
          ],
        },
      };
      const examplePath = process.env[`EXAMPLE_${type}`]!;
      const reactExamplePath = join(dirName(), examplePath);
      const reactPackage = fs.readJSONSync(
        join(dirName(), examplePath, "package.json")
      ) as PackageJSON;
      const hashExample = await hashElement(reactExamplePath, options);
      await createTool().parseAsync([
        "node",
        "cli",
        projectName,
        `-t${type.toLowerCase()}`,
        "--no-github",
      ]);
      const hashResult = await hashElement(
        getProjectPath(projectName),
        options
      );
      const resultPackage = fs.readJSONSync(
        join(getProjectPath(projectName), "package.json")
      ) as PackageJSON;
      const resultReadme = fs.readFileSync(
        join(getProjectPath(projectName), "README.md"),
        "utf-8"
      );
      console.log(resultReadme);
      rimrafSync(getProjectPath(projectName));
      expect(hashResult).toStrictEqual(hashExample);
      expect(resultPackage.dependencies).toStrictEqual(
        reactPackage.dependencies
      );
      expect(resultPackage.devDependencies).toStrictEqual(
        reactPackage.devDependencies
      );
    }
  );
});
