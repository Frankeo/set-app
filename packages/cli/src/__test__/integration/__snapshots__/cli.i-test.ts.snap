// Vitest Snapshot v1

exports[`SetApp cli tool > when specify --help then show help and exit 1`] = `
"Usage: cli [options] <name>

Create and setup node projects

Arguments:
  name                project name

Options:
  -t, --type <value>  Type of node project (choices: \\"react\\", \\"react-redux\\",
                      \\"console\\", default: \\"react\\")
  --desc <value>      Add a description to your new project
  --no-github         Not generate Github public repository
  --no-deploy         Not generate Github-Action for deployment
  -V, --version       output the version number
  -h, --help          display help for command
"
`;

exports[`SetApp cli tool > when specify -h then show help and exit 1`] = `
"Usage: cli [options] <name>

Create and setup node projects

Arguments:
  name                project name

Options:
  -t, --type <value>  Type of node project (choices: \\"react\\", \\"react-redux\\",
                      \\"console\\", default: \\"react\\")
  --desc <value>      Add a description to your new project
  --no-github         Not generate Github public repository
  --no-deploy         Not generate Github-Action for deployment
  -V, --version       output the version number
  -h, --help          display help for command
"
`;
